File: consumer-fixture-langs/scala/src/test/scala/se/deversity/asynctest/fixture/LostUpdateIsReportedTest.scala
package se.deversity.asynctest.fixture

import org.junit.jupiter.api.{AfterAll, BeforeAll}
import se.deversity.asynctest.{AsyncFindings, AsyncTest, AsyncTestContext, FailOn}

/**
 * `@AsyncTest` from Scala 3: eight threads write the same `var` with no lock, and
 * `RaceConditionDetector` reports it.
 *
 * The annotation goes on a `def` in a `class` and JUnit discovers it as-is. The JUnit lifecycle
 * methods live in the companion `object`: Scala 3 emits static forwarders for them, which is
 * what JUnit needs for `@BeforeAll` / `@AfterAll`. Under sbt this needs `sbt-jupiter-interface`
 * to run Jupiter at all; this module is Maven and Gradle like the rest of the repository.
 *
 * One Scala-specific fact worth knowing: a `var` compiles to accessors named `counter()` and
 * `counter_$eq(int)`, not `getCounter`/`setCounter`, so the agent's default accessor weaving
 * does not see it. The explicit hook below makes the write observable regardless;
 * `-Dasynctest.agent=fields=true` weaves the field instructions themselves.
 */
object LostUpdateIsReportedTest {
  private var findings: AsyncFindings = scala.compiletime.uninitialized

  @BeforeAll
  def collect(): Unit =
    findings = AsyncFindings.collect()

  @AfterAll
  def theRaceWasReported(): Unit =
    try findings.assertReported("RaceConditionDetector")
    finally findings.close()
}

class LostUpdateIsReportedTest {

  private var counter = 0

  @AsyncTest(threads = 8, invocations = 200, failOn = FailOn.NONE, licenseMockMode = true)
  def unguardedWritesFromEightThreads(): Unit = {
    val ctx = AsyncTestContext.get()
    if (ctx != null) ctx.sharedRaceConditionDetector().recordFieldWrite(this, "counter")
    counter += 1
  }
}

File: consumer-fixture-langs/kotlin/src/test/kotlin/se/deversity/asynctest/fixture/LostUpdateIsReportedTest.kt
package se.deversity.asynctest.fixture

import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import se.deversity.asynctest.AsyncFindings
import se.deversity.asynctest.AsyncTest
import se.deversity.asynctest.AsyncTestContext
import se.deversity.asynctest.FailOn

/**
 * `@AsyncTest` from Kotlin, the direction that must fire: eight threads write the same field with
 * no lock, and `RaceConditionDetector` reports it.
 *
 * Nothing here is Kotlin-specific beyond syntax. `@AsyncTest` is a JUnit 5 `@TestTemplate`, its
 * attributes are named arguments, and the JUnit lifecycle methods live in a `companion object`
 * with `@JvmStatic` because JUnit needs them static.
 *
 * The finding is asserted in `@AfterAll`, not in the body: detectors analyse after the last round,
 * so a finding cannot be observed from inside the test. `failOn = FailOn.NONE` keeps the report
 * from failing the test itself, so the assertion below is the only thing that can go red.
 */
class LostUpdateIsReportedTest {

    companion object {
        private lateinit var findings: AsyncFindings

        @JvmStatic
        @BeforeAll
        fun collect() {
            findings = AsyncFindings.collect()
        }

        @JvmStatic
        @AfterAll
        fun theRaceWasReported() {
            try {
                findings.assertReported("RaceConditionDetector")
            } finally {
                findings.close()
            }
        }
    }

    private var counter = 0

    @AsyncTest(threads = 8, invocations = 200, failOn = FailOn.NONE, licenseMockMode = true)
    fun unguardedWritesFromEightThreads() {
        // The hook is what makes the write observable to the detector without the agent; a
        // consumer using -Dasynctest.agent=fields=true gets the same event from the weaver.
        AsyncTestContext.get()?.sharedRaceConditionDetector()?.recordFieldWrite(this, "counter")
        counter++
    }
}
